File: extension/tests/sync/historyMaterializer.test.ts
import { describe, expect, test } from "bun:test";
import {
  applyTag,
  authorizeDestructiveHistory,
  dailyTotals,
  materializeHistory,
  type HistoryBlock,
} from "../../src/sync/domain/history";

function block(blockId: string, elapsedMillis: number): HistoryBlock {
  return {
    blockId,
    phaseId: `phase-${blockId}`,
    startedAtEpochMillis: 1_755_734_400_000,
    elapsedMillis,
    outcome: "COMPLETED",
    tagId: "tag-work",
    authoredTagName: "Work",
    localDate: "2026-08-21",
  };
}

describe("identified History and stable Session tags", () => {
  test("retains correction and tombstone alternatives until settlement", () => {
    const original = block("stable-a", 60_000);
    const corrected = { ...original, elapsedMillis: 90_000 };
    const facts = [
      { kind: "CREATE" as const, factId: "create", blockId: original.blockId, block: original },
      { kind: "CORRECT" as const, factId: "correct", blockId: original.blockId, replacement: corrected },
      { kind: "TOMBSTONE" as const, factId: "delete", blockId: original.blockId },
    ];
    expect(materializeHistory(facts).conflicts).toContain(original.blockId);
    const settled = materializeHistory([...facts, {
      kind: "SETTLE" as const,
      factId: "settle",
      blockId: original.blockId,
      selectedFactIds: new Set(["create", "correct"]),
    }]);
    expect(settled.visible.get(original.blockId)?.elapsedMillis).toBe(90_000);
    expect(settled.alternatives.get(original.blockId)).toHaveLength(4);
  });

  test("rebuilds totals and never infers deletion from an absent local row", () => {
    const first = block("stable-a", 60_000);
    const second = { ...block("stable-b", 30_000), outcome: "PARTIAL" as const };
    const projection = materializeHistory([
      { kind: "CREATE" as const, factId: "a", blockId: first.blockId, block: first },
      { kind: "CREATE" as const, factId: "b", blockId: second.blockId, block: second },
    ]);
    expect([...projection.visible.keys()]).toEqual(["stable-a", "stable-b"]);
    expect(dailyTotals(projection).get("2026-08-21")).toEqual({ elapsedMillis: 90_000, completed: 1 });
  });

  test("preserves Tag IDs and independently confirms broad destructive scope", () => {
    const work = { tagId: "tag-work", name: "Work", paletteSlot: 0, archived: false, mergedInto: null };
    const study = { tagId: "tag-study", name: "Study", paletteSlot: 1, archived: false, mergedInto: null };
    const renamed = applyTag(new Map([[work.tagId, work], [study.tagId, study]]), { ...study, name: "Deep work" }, work.tagId, study.tagId);
    const archived = applyTag(renamed.tags, { ...study, name: "Deep work", archived: true }, work.tagId, study.tagId);
    expect(archived.tags.get(study.tagId)?.name).toBe("Deep work");
    expect(archived.defaultTagId).toBe(work.tagId);
    const broad = new Set(Array.from({ length: 10 }, (_, index) => `${index}`));
    expect(authorizeDestructiveHistory(broad, new Set())).toBeFalse();
    expect(authorizeDestructiveHistory(broad, broad)).toBeTrue();
  });
});

File: app/src/test/java/com/pomo/sync/domain/HistoryMaterializerTest.kt
package com.pomo.sync.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

public class HistoryMaterializerTest {
    @Test
    public fun concurrentCorrectionAndTombstoneRemainUntilExplicitSettlement() {
        val original = block("block-1", 1_500, "tag-work", "Work")
        val corrected = original.copy(elapsedMillis = 1_800)
        val facts =
            listOf(
                HistoryFact.Create("create", original),
                HistoryFact.Correct("correct", original.blockId, corrected),
                HistoryFact.Tombstone("delete", original.blockId),
            )
        val conflicted = HistoryMaterializer().materialize(facts)
        assertTrue(original.blockId in conflicted.conflicts)
        assertEquals(3, conflicted.alternatives.getValue(original.blockId).size)

        val settled =
            HistoryMaterializer().materialize(
                facts + HistoryFact.Settle("settle", original.blockId, setOf("create", "correct")),
            )
        assertEquals(1_800L, settled.visible.getValue(original.blockId).elapsedMillis)
        assertFalse(original.blockId in settled.conflicts)
    }

    @Test
    public fun missingHistoryDoesNotDeleteAndDerivedTotalsRebuildFromVisibleBlocks() {
        val materializer = HistoryMaterializer()
        val first = block("stable-a", 60_000, "tag-work", "Work")
        val second = block("stable-b", 30_000, "tag-study", "Study").copy(outcome = HistoryOutcome.PARTIAL)
        val projection = materializer.materialize(listOf(HistoryFact.Create("a", first), HistoryFact.Create("b", second)))
        assertEquals(90_000L to 1, materializer.dailyTotals(projection).getValue("2026-08-21"))
        assertEquals(setOf("stable-a", "stable-b"), projection.visible.keys)
    }

    @Test
    public fun settlementSelectingMultipleCorrectionsRemainsConflicted() {
        val original = block("block-2", 1_500, "tag-work", "Work")
        val facts =
            listOf(
                HistoryFact.Create("create", original),
                HistoryFact.Correct("correct-a", original.blockId, original.copy(elapsedMillis = 1_600)),
                HistoryFact.Correct("correct-b", original.blockId, original.copy(elapsedMillis = 1_700)),
                HistoryFact.Settle("settle", original.blockId, setOf("create", "correct-a", "correct-b")),
            )

        val materialized = HistoryMaterializer().materialize(facts)

        assertTrue(original.blockId in materialized.conflicts)
        assertTrue(original.blockId !in materialized.visible)
    }

    @Test
    public fun settlementSelectingForeignBlockFactRemainsConflicted() {
        val local = block("block-local", 1_500, "tag-work", "Work")
        val foreign = block("block-foreign", 2_000, "tag-study", "Study")
        val facts =
            listOf(
                HistoryFact.Create("create-local", local),
                HistoryFact.Correct("correct-local", local.blockId, local.copy(elapsedMillis = 1_800)),
                HistoryFact.Create("create-foreign", foreign),
                HistoryFact.Settle("settle-local", local.blockId, setOf("create-foreign")),
            )

        val materialized = HistoryMaterializer().materialize(facts)

        assertTrue(local.blockId in materialized.conflicts)
        assertTrue(local.blockId !in materialized.visible)
        assertEquals(2_000L, materialized.visible.getValue(foreign.blockId).elapsedMillis)
    }

    @Test
    public fun tagIdentitySurvivesRenameArchiveRestoreAndMerge() {
        val work = SessionTag("tag-work", "Work", 0)
        val study = SessionTag("tag-study", "Study", 1)
        val materializer = TagMaterializer(work.tagId)
        var state = mapOf(work.tagId to work, study.tagId to study)
        var default = study.tagId
        materializer.apply(state, study.copy(name = "Deep work"), default).also { (next, selected) ->
            state = next
            default = selected
        }
        materializer.apply(state, state.getValue(study.tagId).copy(archived = true), default).also { (next, selected) ->
            state = next
            default = selected
        }
        assertEquals(work.tagId, default)
        assertEquals("Deep work", state.getValue(study.tagId).name)
        assertTrue(DestructiveHistoryGuard.authorize(setOf("a"), emptySet()))
        assertFalse(DestructiveHistoryGuard.authorize((1..10).map { "$it" }.toSet(), emptySet()))
    }

    private fun block(
        id: String,
        elapsed: Long,
        tagId: String,
        tagName: String,
    ): HistoryBlock =
        HistoryBlock(
            id,
            "phase-$id",
            1_755_734_400_000,
            elapsed,
            HistoryOutcome.COMPLETED,
            tagId,
            tagName,
            "2026-08-21",
        )
}
